extkeys/mnemonic.go

```diff
@@ -216,17 +232,61 @@ func (m *Mnemonic) ValidMnemonic(mnemonic string, language Language) bool {
 
 	// The number of words should be 12, 15, 18, 21 or 24
 	if numOfWords%3 != 0 || numOfWords < 12 || numOfWords > 24 {
-		return false
+		return errors.New("mnemonic contains an invalid number of words")
 	}
 
+	// Create reverse lookup map for dictionary
+	wordMap := map[string]int{}
+	for i, v := range wordList {
+		wordMap[v] = i
+	}
+
+	b := big.NewInt(0)
+
 	// Check if all words belong in the wordlist
 	for i := 0; i < numOfWords; i++ {
-		if !contains(wordList, words[i]) {
-			return false
+
+		wordIndex, found := wordMap[words[i]]
```

Comment:
The pattern is usually

```go
wordIndex, ok := wordmap[words[i]]
if !ok {
	return errors.New("not all words are found in dictionary")
}
```

Reply:
Understood, will be updated in next commit.